### ISSUE TITLE
Holy Paladin's Illumination talent is refunding the incorrect mana cost when casting spells in quick succession

### ISSUE THREAD

**Issue by MrSmithers** [Wall-core/Everlook-Bugtracker#722]
When a Holy Paladin has taken the Illumination talent and critically heals, 100% of the base mana cost is refunded. This works a treat, but an incorrect mana cost is refunded when the refund comes after you've begun casting your next spell. The refunded mana is instead calculated based on your next cast.

Here's an example of using Holy Light rank 7 (465 base mana cost) followed by Summon Warhorse (100 base mana cost)
![Capture](https://user-images.githubusercontent.com/13151207/204886704-5ac31b4a-4885-4b3c-8ebd-13f2d09bcce4.PNG)

Here's a second example of a Flash of Light rank 4 (90 base mana cost) followed by Summon Warhorse (100 base mana cost), demonstrating this has the potential to be "cheesed" to gain more mana back than you should:
![Illum2](https://user-images.githubusercontent.com/13151207/204886705-a1f53164-828f-4246-9249-5796450ee326.PNG)

Here's a third example showing this happens even without Diving Favor, instead just on a regular crit. Here I have crit with Flash of Light rank 4 (90 base mana cost), followed by casting Seal of Light rank 2 (140 base mana cost):
![Illum3](https://user-images.githubusercontent.com/13151207/204888000-d8fe7545-9ec6-48e5-a1b3-07484192480c.PNG)

Final interesting nugget of info this bug, using Cleanse after your crit heal results in a mana refund of 0! 
![Illum4](https://user-images.githubusercontent.com/13151207/204889584-13e4a80f-70b1-40ca-babb-4eff12738fcb.PNG)

Edit1: Included image of cleanse 0 mana refund

**Comment by MrSmithers** [comment]
Thanks to @Darg for suggesting this. Cheese taken to the extreme, you can spam cast rank 1 Holy Light followed by max rank Holy Light. Crits will restore the mana cost of the max rank ability:
![Illum5](https://user-images.githubusercontent.com/13151207/204916959-0bbfa5e4-ee9e-4460-90a4-8b27f27f0901.PNG)

**Comment by Godmatik** [comment]
The same batching issues occur with elemental mastery, just different applications

**Comment by MrSmithers** [comment]
@Godmatik Seems almost the reverse problem, in fact! Elemental Mastery is consumed on a spell hit, rather than spell cast, and illumination is triggered on spell cast rather than spell hit! These issues do seem to be related, but I think this one has potential to be exploited in a raid environment for the benefit of the player, whereas the EM bug is just bad for the player!

**Comment by Gamemechanicwow** [comment]
> The same batching issues occur with elemental mastery, just different applications

It's not related to batching.

**Comment by MrSmithers** [comment]
> > The same batching issues occur with elemental mastery, just different applications
> 
> It's not related to batching.

Illumination appears to restore mana on server tick. I can't comment on the EM issue.

**Comment by Gamemechanicwow** [comment]
> Illumination appears to restore mana on server tick. I can't comment on the EM issue.

Nope, it's a different issue where it returns mana from the current spell being cast instead of the spell causing the proc. I submitted a fix for review.
